TraktorFeature: Fixes to Traktor collection importer

DidierMalenfant opened the pull request:
First PR... be kind :)

Hopefully I'm not misunderstanding some of the intents behind the original code I'm modifying here. If I'm missing some corner cases that this was written done for, let me know.

I tracked down the change that introduced the std::move to a 2022 commit for example. The intent was probably to fix a warning that came with QSqlQuery not being copyable anymore but I'm not sure how the fix ever worked so I'm questioning my understanding of this.

DidierMalenfant committed:
Fixed a crash caused by local variables being moved and then reused

If the Traktor collection contains more than one playlist then query_insert_to_playlists and query_insert_to_playlist_tracks cannot be moved more than once and cause an undefined behavior.

DidierMalenfant committed:
Fixed incorrect folder hierarchy importing

If a folder contained a node that was something else that a folder or a playlist (SMARTLIST for example) then the node would not be added to the current_path but it would be removed from current_path when the node closed. This would cause the folder hierarchy to get out of sync and display incorrectly.

This now processes every node but only adds playlists and non-empty folder to the item tree.

DidierMalenfant committed:
Minor whitespace fixes

DidierMalenfant committed:
Some locals could be made const

daschuer commented on src/library/traktor/traktorfeature.cpp:
```diff
@@ -414,48 +414,43 @@ void TraktorFeature::parseTrack(QXmlStreamReader &xml, QSqlQuery &query) {
+    QSet<QString> playlists;
 
-    QString delimiter = "-->";
+    const QString delimiter = "-->";
```
this avoids runtime Utf8 to Utf16 conversion: 
```suggestion
    const QString delimiter = QStringLiteral("-->");
```

daschuer reviewed (changes requested):
Thank you for finding and fixing this severe bug. 

The current PR should already work. I have added some hints to speed up the code. 
This is not really a performance bottle neck, just treat it as knowledge exchange.
It is also not your fault, the code was in that shape before. 
Here the Boy Scouts rule applies: "try to leave the camp ground a little better than you found it". I have tried to also explain the reasons. Hopefully it helps, since I am not sure how familiar you are with Qt.

DidierMalenfant committed:
Fixed pre-commit coding style errors

DidierMalenfant committed:
Fixed minor casing error and typos in comments

DidierMalenfant committed:
Optimized handling of some string literals

QStringLiteral avoids deep copy of the string. QStringView provides a read-only subset of the QString API.

DidierMalenfant committed:
Renamed QLatin1String to QLatin1StringView

QLatin1StringView is the new recommended name for QLatin1String. Using this name makes it clearer that this is a thin-wrapper around a string and not a full mutable string object.

DidierMalenfant commented on src/library/traktor/traktorfeature.cpp:
```diff
@@ -414,48 +414,43 @@ void TraktorFeature::parseTrack(QXmlStreamReader &xml, QSqlQuery &query) {
+                                  "VALUES (:name)");
+
+                    QSqlQuery query_insert_to_playlist_tracks(m_database);
+                    query_insert_to_playlist_tracks.prepare(
```
Gotcha. When I change the method arguments to be pointer, should I change the names to `pQueryInsertIntoPlaylist` and `pQueryInsertIntoPlaylistTracks` instead of the snake_case?

I've seen that elsewhere in the code to de-notate pointers.

daschuer replied:
I would prefer that. But we have no strict policy.

DidierMalenfant replied:
One thing about this is that I don't think there is a way to un-bind values in queries.

Probably not a huge deal in the grand scheme of things but as it stands, if someone modifies parsePlaylistEntries() and doesn't bind all the values for some reason, the bound values from the previous call will still be there even after calling finish() I believe.

DidierMalenfant committed:
Changed QLatin1StringView into QStringLiteral

Saves a conversion from UTF8 to Utf16.

DidierMalenfant committed:
Converted a literal string into a QStringLiteral

DidierMalenfant committed:
Fixed a pre-commit coding style error

DidierMalenfant committed:
Replaced a QString constructor call with toString() method call

QString::QString(QStringView sv) was only introduced in Qt6.8 so this should work better with older versions.

DidierMalenfant committed:
TraktorFeature: parsePlaylistEntries() now takes pointers to QSqlQuery

This allows us to reuse the same queries each time parsePlaylistEntries() is called and only bind new values each time.

DidierMalenfant commented:
Thank you for your patience with me here. I'm getting the swing of things slowly :)

I did work on a Qt app in the past but It's always good to refresh my memory with the proper best practices. I'm always here to learn something.

DidierMalenfant commented:
Found another issue in importing a playlist or a crate. They are stores in the db as paths like '-->Folder-->Folder-->Playlist' but the importer only looks for 'Playlist' and fails to find it.

I should probably add a fix to that as part of this PR.

DidierMalenfant committed:
Fixed line length formatting

daschuer reviewed (approved):
LGTM, thank you. 
I have not tested it because of the absence of a Traktor installation. But I rely on your own tests .
